_posts/2021-02-19-26-rln-relay.md

```diff
@@ -0,0 +1,195 @@
+
+Finally, the peer generates a zero-knowledge proof  `zkProof`  asserting the membership of the peer in the group and the correctness of the attached secret share (`shareX`, `shareY`) and `nullifier`. In order to generate a valid proof, the peer needs to have its `sk`, its authentication path, the tree root, epoch, and obviously the content of the message.
+
+**Privacy Hint:** Note that the authentication path of each peer depends on the recent list of members (hence changes when new peers register or leave). As such, it is recommended (and necessary for privacy/anonymity) that the publisher updates her authentication path based on the latest status of the group and attempts the proof using the updated version.
+
+An overview of the publishing procedure is provided in Figure 2.
+
+## Routing
+
+Upon the receipt of a message, the routing peer needs to decide whether to route it or not. This decision relies on the following factors:
+1- If the epoch value attached to the message has a non-reasonable gap with the routing peer's current epoch then the message must be dropped (this is to prevent a newly registered peer spamming the system by messaging for all the past epochs). 
+2- The message MUST contain valid proof that gets verified by the routing peer. 
+If the preceding checks are passed successfully, then the message is Relayed. In case of an invalid proof, the message is dropped. If spamming is detected, the publishing peer gets slashed (see [Spam Detection and Slashing](#Spam-Detection-and-Slashing)). 
+
+An overview of the routing procedure is provided in Figure 2.
+
+### Spam Detection and Slashing
+In order to enable local spam detection and slashing, routing peers MUST record the `nullifier`, `shareX`, and `shareY` of any incoming message conditioned that it is not spam and has valid proof. To do so, the peer should follow the following steps. 
+1. The routing peer first verifies the `zkProof` and drops the message if not verified. 
+2. Otherwise, it checks whether a message with an identical `nullifier` has already been Relayed. 
+   1. If such message exists and its `shareX` and `shareY` components are different from the incoming message, then slashing takes place (if the `shareX` and `shareY` fields of the previously Relayed message is identical to the incoming message, then the message is a duplicate and shall be dropped).
+   2. If none found, then the message gets Relayed.
+
+An overview of the slashing procedure is provided in Figure 2.
+
+<p align="center">
+    <img src="../assets/img/RLN-Relay/RLN-message-verification.png" />
+    <br />
+    Figure 2: Publishing, Routing and Slashing workflow.
+</p>
+
+
+# Feasibility and Open Issues
```

Comment:
Maybe brief intro "We've come a long way since a year ago, blockers resolved, now we have implemented it e2e and learned a lot, identified more issues etc etc. Still some issues blocking getting to production..."